hazelcast/src/main/java/com/hazelcast/datalink/impl/DataLinkConsistencyChecker.java

```diff
@@ -0,0 +1,110 @@
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.datalink.impl.DataLinkServiceImpl.DataLinkSource;
+import com.hazelcast.datalink.impl.DataLinkServiceImpl.DataLinkEntry;
+import com.hazelcast.jet.impl.JetServiceBackend;
+import com.hazelcast.map.IMap;
+import com.hazelcast.spi.impl.NodeEngine;
+import com.hazelcast.sql.impl.QueryUtils;
+import com.hazelcast.sql.impl.schema.datalink.DataLinkCatalogEntry;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DataLinkConsistencyChecker {
+    private final HazelcastInstance hazelcastInstance;
+    private final DataLinkServiceImpl dataLinkService;
+    // sqlCatalog supposed to be initialized lazily
+    private IMap<Object, Object> sqlCatalog;
+
+    private boolean initialized;
+
+    public DataLinkConsistencyChecker(HazelcastInstance instance, NodeEngine nodeEngine) {
+        this.hazelcastInstance = instance;
+        this.dataLinkService = (DataLinkServiceImpl) nodeEngine.getDataLinkService();
+    }
+
+    /**
+     * Lazily initialize sqlCatalog.
+     * The only reason of this method existing is
+     * to prevent eager creation of SQL catalog IMap.
+     */
+    public void init() {
+        sqlCatalog = hazelcastInstance.getMap(JetServiceBackend.SQL_CATALOG_MAP_NAME);
+        initialized = true;
+    }
+
+    /**
+     * Removes & alters all outdated data links in __sql.catalog
+     * and adds all missed data links to __sql.catalog
+     */
+    public void check() {
+        if (!initialized) {
+            return;
+        }
+
+        List<Map.Entry<String, DataLinkEntry>> sqlEntries = dataLinkService.getDataLinks()
+                .entrySet()
+                .stream()
+                .filter(en -> en.getValue().source == DataLinkSource.SQL)
+                .collect(Collectors.toList());
+
+        for (Map.Entry<String, DataLinkEntry> entry : sqlEntries) {
+            Object catalogValue = sqlCatalog.get(QueryUtils.wrapDataLinkKey(entry.getKey()));
+
+            // Data link is absent in sql catalog -> remove it from data link service.
+            if (catalogValue == null) {
+                dataLinkService.removeDataLink(entry.getKey());
+            } else {
+                // Try to alter outdated data links
+                DataLinkCatalogEntry dl = (DataLinkCatalogEntry) catalogValue;
+                if (!dataLinksAreEqual(entry.getValue(), dl)) {
+                    dataLinkService.replaceSqlDataLink(dl.getName(), dl.getType(), dl.getOptions());
+                }
+            }
+        }
+
+        // Data link is not present in service -> add it
+        for (Map.Entry<Object, Object> entry : sqlCatalog.entrySet()) {
+            if (!(entry.getValue() instanceof DataLinkCatalogEntry)) {
+                continue;
+            }
+            DataLinkCatalogEntry dl = (DataLinkCatalogEntry) entry.getValue();
+            if (!dataLinkService.existsSqlDataLink(dl.getName())) {
```

Comment:
you don't have to check if the link exists, `dataLinkService.replaceSqlDataLink` will take care of it. BTW, now it might be simpler to reshape the loops (less ifs):
```java
for (Map.Entry<Object, Object> entry : sqlCatalog.entrySet()) {
    DataLinkCatalogEntry dl = (DataLinkCatalogEntry) entry.getValue();
    dataLinkService.replaceSqlDataLink(dl.getName(), dl.getType(), dl.getOptions());
}

for (Map.Entry<String, DataLinkEntry> entry : sqlEntries) {
            Object catalogValue = sqlCatalog.get(QueryUtils.wrapDataLinkKey(entry.getKey()));

            // Data link is absent in sql catalog -> remove it from data link service.
            if (catalogValue == null) {
                dataLinkService.removeDataLink(entry.getKey());
            }
}
```

Reply:
Yes, but current version is straightforward and simple (IMO). Your variant contains implicit knowledge that `replaceSqlDataLink` would take care of it, but it's not obvious. I'd preserve the original solution.